File: 328/kcl.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        ListNode *o=NULL;
        ListNode *e=NULL;
        ListNode *oh=NULL;
        ListNode *eh=NULL;
        ListNode *tmp=NULL;
        i=0;
        while(head){
            i++;
            tmp=head->next;
            head->next=NULL;
            if(i%2==0){
                // even
                if(e==NULL){
                    e=head;
                    eh=e;
                }else{
                    e->next=head;
                    e=e->next;
                }
            }else{
                // odd
                if(o==NULL){
                    o=head;
                    oh=head;
                }else{
                    o->next=head;
                    o=o->next;
                }
            }
            head=tmp;
        }
        if(o){
            o->next=eh;
            return oh;
        }
        return eh;
    }
};

File: 234/kcl.go
package main

import "fmt"

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
type ListNode struct {
    Val int
    Next *ListNode
}

func isPalindrome(head *ListNode) bool {
	if head == nil{
		return true
	}
	// find middle position
	slow := head
	fast := head.Next
	for fast != nil && fast.Next != nil{
		slow = slow.Next
		fast = fast.Next.Next
	}

	sec := slow.Next
	slow.Next = nil
	fmt.Println(sec)

	// reverse 
	sec = reverseList(sec)
	for head != nil && sec != nil{
		if head.Val != sec.Val{
			return false
		}
		head = head.Next
		sec = sec.Next
	}
	return true
}

func reverseList(head *ListNode) *ListNode{
	if head == nil{
		return head
	}
	var left *ListNode
	for head != nil{
		tmp := head.Next
		head.Next = left
		left = head
		head = tmp
	}
	return left
}

func main() {
	/*
	n5 := &ListNode{5, nil}
	n4 := &ListNode{4, n5}
	n3 := &ListNode{3, n4}
*/
	n2 := &ListNode{0, nil}
	n1 := &ListNode{0, n2}
	fmt.Println(isPalindrome(n1))
}
